agents/agents-core/src/commonMain/kotlin/ai/koog/agents/core/agent/ActAIAgent.kt

```diff
@@ -0,0 +1,211 @@
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import kotlinx.datetime.Clock
+import kotlin.uuid.ExperimentalUuidApi
+import kotlin.uuid.Uuid
+
+/**
+ * Represents the core AI agent for processing input and generating output using
+ * a defined configuration, toolset, and prompt execution pipeline.
+ *
+ * @param Input The type of input data expected by the agent.
+ * @param Output The type of output data produced by the agent.
+ * @param id The unique identifier for the agent instance.
+ * @param promptExecutor The executor responsible for processing prompts and interacting with language models.
+ * @param agentConfig The configuration for the agent, including the prompt structure and execution parameters.
+ * @param toolRegistry The registry of tools available for the agent. Defaults to an empty registry if not specified.
+ */
+@ExperimentalUuidApi
+public class ActAIAgent<Input, Output>(
+    public val promptExecutor: PromptExecutor,
+    public val agentConfig: AIAgentConfigBase,
+    override val id: String = Uuid.random().toString(),
+    public val toolRegistry: ToolRegistry = ToolRegistry.EMPTY,
+    public val strategy: AIAgentLoopStrategy<Input, Output>,
+    public val clock: Clock = Clock.System,
+    public val featureContext: FeatureContext.() -> Unit = {}
+) : AIAgent<Input, Output> {
+
+    private companion object {
+        private val logger = KotlinLogging.logger {}
+    }
+
+    private val pipeline = AIAgentNonGraphPipeline()
+
+    private val environment = GenericAgentEnvironment(
+        id,
+        strategy.name,
+        logger,
+        toolRegistry,
+        pipeline = pipeline
+    )
+
+    /**
+     * Represents a context for managing and configuring features in an AI agent.
+     * Provides functionality to install and configure features into a specific instance of an AI agent.
+     */
+    public class FeatureContext internal constructor(private val agent: ActAIAgent<*, *>) {
+        /**
+         * Installs and configures a feature into the current AI agent context.
+         *
+         * @param feature the feature to be added, defined by an implementation of [AIAgentFeature], which provides specific functionality
+         * @param configure an optional lambda to customize the configuration of the feature, where the provided [Config] can be modified
+         */
+        public fun <Config : FeatureConfig, Feature : Any> install(
+            feature: AIAgentNonGraphFeature<Config, Feature>,
+            configure: Config.() -> Unit = {}
+        ) {
+            agent.install(feature, configure)
+        }
+    }
+
+    private var isRunning = false
+
+    private val runningMutex = Mutex()
+
+    private fun <Config : FeatureConfig, Feature : Any> install(
+        feature: AIAgentNonGraphFeature<Config, Feature>,
+        configure: Config.() -> Unit
+    ) {
+        pipeline.install(feature, configure)
+    }
+
```

Comment:
```suggestion
    init {
        FeatureContext(this).featureContext()
    }
```

Reply:
I might be missing something important/intentional but the features work for me when I add this :)

side note, I really like these interfaces!

Reply:
@CaelumF Good catch, thank you!